SONARJAVA-5425 Create rule S7466: Unnamed variable declarations should use var identifier (#4860)
SonarSource/rspec: rules/S7466/java/rule.adoc

== Why is this an issue?

Java 22 introduced unnamed variable pattern, `_`.

> This feature alleviates the need to name variables that are syntactically required but otherwise unused.
> Moreover, it clearly indicates the intent not to use the variable.

Reviewer: Can we simplify the vocabulary here?
```suggestion
This feature removes the need to name variables that are syntactically required but otherwise unused.
Moreover, it clearly indicates the intent not to use the variable.
```

To further minimize code clutter, unnamed variables should use the `var` pattern rather than explicit type declarations.

> This will also improve maintainability should the type change in future.

Reviewer: Can we connect the two ideas in a smoother way before making a recommendation?

Reviewer 2: Done.

== How to fix it

> Replace type name with `var` pattern.

Reviewer: ```suggestion
Replace the local variable type with `var`.
```

=== Code examples

==== Noncompliant code example

[source,java,diff-id=1,diff-type=noncompliant]
----

> for(String _: myIterable) { // Noncompliant

Reviewer 2: Done.

  // ...
}

> try(Resource _ = new Resource()) { // Noncompliant

Reviewer 2: Done.

== Why is this an issue?

Java 22 introduced unnamed variable pattern, `_`.
This feature removes the need to name variables that are syntactically required but otherwise unused.
Moreover, it clearly indicates the intent not to use the variable.

To further minimize code clutter, unnamed variables should use the `var` pattern rather than explicit type declarations.

> Removing unnecessary information is done to make code more readable, but it brings an additional benefit in maintainability: should the type change in future, the code can remain the same.

Reviewer: Let's try to shorten this and connect it to the sentence above
```suggestion
In addition to minimizing clutter, removing the type makes the code easier to maintains.
Indeed, if the type was to change in future, the code would remain the same.
```
